---
title: "EDA for CRC patient RNAseq data"
output:
  html_document:
    df_print: paged
date: "2023-08-03"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load the packages and data

```{r load packages, message=FALSE, warning=FALSE}
library(SummarizedExperiment)
library(DESeq2)
library(dplyr)
library(stringr)
library(stats)
library(limma)
library(edgeR)
```

```{r load data}
se <- readRDS('../data/processed_AB_20230627/salmon.merged.gene_counts.rds')
```

### SE object has three compartments:

1.  **assay** - the matrix of counts;
2.  **rowRanges** - contain info about genomic ranges;
3.  **colData** - info about samples

![components of SummarizedExperiment object](images/SEsummarize.svg)

```{r check assay first}
head(assay(se, "counts")) # estimated for each gene and sample
head(assay(se, 'abundance')) # estimated transcripts abundance in TPM
head(colSums(assay(se))) #can also check the sum of all counts for all samples
```

> p.s. note that the raw counts are not integers, so cannot proceed with DESeq2 as it requires integers

```{r now check the colData}
colData(se) #maybe can update it later
```

```{r check rowdata, message=TRUE}
head(rowData(se))
```

```{r about $ - it operates on colData() columns}
se[,se$names == "S1_6310_Poly"] 
```

### Update colData for more info

```{r to include site and sample ID}
NEWcolData <- data.frame(colData(se))
NEWcolData <- NEWcolData %>% 
  mutate(site = case_when(
    endsWith(names, "Nor") ~ "Norm",
    endsWith(names, "Norm") ~ "Norm",
    endsWith(names, "Pol") ~ "Poly",
    endsWith(names, "Poly") ~ "Poly")) %>%
  mutate(patient = str_split_fixed(names, fixed("_"), 3)[,2])
NEWcolData[,3:4]
```

```{r update the colData to se}
se$tissue <-NEWcolData[,3] # add tissue info
se$patient <- NEWcolData[,4] #add patients info
colData(se)
```

## Generate DGE object

```{r}
library(limma)
library(edgeR)
```

```{r}
dge <- DGEList(counts = assay(se,'counts'), 
               samples = colData(se),
               genes = rowData(se)[,2])
dim(dge)
```

```{r remove low expressed genes}
table(rowSums(dge$counts==0)==42)
```

```{r}
table(rowSums(dge$counts==0)==42)
```

`{sprintf("There are %1$d genes that are NOT detected in any samples, this account for %3$.2f %% out of %2$d all genes from the raw data",}         nrow(dge$counts[rowSums(dge$counts==0)==42, ]),          dim(dge$counts)[1],          nrow(dge$counts[rowSums(dge$counts==0)==42, ])/dim(dge$counts)[1]*100)`

### Adding gene annotation

```{r, message=FALSE, warning=F}
# use a Ensemble based db
library(EnsDb.Hsapiens.v86)
# usage here https://bioconductor.org/packages/release/bioc/vignettes/ensembldb/inst/doc/ensembldb.html
```

```{r}
# save the db locally
edb <- EnsDb.Hsapiens.v86
```

```{r}
supportedFilters(edb)
```

```{r}
keytypes(edb)
# will need the "EntrezFilter" to convert ENSEMBL id to NCBIs
```

```{r, warning=FALSE, message=FALSE}
# update gene annotation
annot <- select(edb, 
                rownames(dge$genes), 
                keytype = "GENEID", 
                columns = c("ENTREZID","GENEID"))

dim(annot)[1]
head(annot)
```

```{r calculate cpm for raw data - notes for myself}
cpm.raw <- cpm(dge)
lcpm.raw <- cpm(dge, log=T)

# understand log2CPM
L <- mean(dge$samples$lib.size)*1e-6
M <- median(dge$samples$lib.size)*1e-6
# this is normally distributed??
# cpm function adds an offset to the CPM values before converting to log2
# by default, this offset is 2/L - where 2 is prior count and L is average lib size in millions
# so logCPM value is related to the CPM valuesa by log2(CPM+2/L)
# hence the smallest value would be logs(0+2/39.33607) = -4.297781

# now lets validate this is true
summary(lcpm.raw)[,1:5]
```

```         
```

```{r}
plot(table(rowSums(dge$counts==0)))
title(xlab = "Number of samples with unexpressed genes across all samples",
      ylab = "Number of genes")
```

## Filtering low expressed genes

```{r}
keep.exprs <- filterByExpr(dge, group = dge$samples$patient)
dge.filter <- dge[keep.exprs,, keep.lib.sizes=F]
dim(dge.filter)
dim(dge)

sprintf('There are %1$d genes detected in the raw data, of which, %2$d is detected in 0 samples.After filtering, %3$d were retained.', 
        dim(dge)[1], 
        nrow(dge$counts[rowSums(dge$counts==0)==42, ]),
        dim(dge.filter)[1]
)
```

```{r plot log-cpm before and after filtering, message=FALSE, warning=FALSE}
library(RColorBrewer)
lcpm.cutoff <- log2(10/M + 2/L) # set the cutoff level
nsamples <- ncol(dge)
col <- brewer.pal(nsamples, 'Paired')

par(mfrow=c(1,2)) # make a 1X2 subplot

# plot the raw data
plot(density(lcpm.raw[,1]), col=col[1], lwd=2, ylim=c(0,0.7), las=2, main="", xlab="")
title(main="A. Raw data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm.raw[,i])
  lines(den$x, den$y, col=col[i], lwd=2)}

# plot the filtered data
lcpm.filter <- cpm(dge.filter, log=TRUE)
plot(density(lcpm.filter[,1]), col=col[1], lwd=2, ylim=c(0,0.3), las=2, main="", xlab="")
title(main="B. Filtered data", xlab="Log-cpm")
abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(lcpm.filter[,i])
  lines(den$x, den$y, col=col[i], lwd=2)} 
legend("topright", dge$samples$patient, text.col=col, bty="n") #add the legend
```

```{r}
table(rowSums(dge.filter$counts==0)>38) # check that genes unexpressed in most cells are removed
plot(table(rowSums(dge.filter$counts==0)))
title(xlab = "Number of samples with unexpressed genes across all samples",
      ylab = "Number of genes")
```

## Normalize gene expression

```{r set conditions}
dge.filter <- calcNormFactors(dge.filter, method = 'TMM')
# this stores the normalization factors in the dge object
dge.filter$samples$norm.factors
```

```{r}
par(mfrow=c(1,2), cex=0.6, mar=c(7,2,2,2))
boxplot(lcpm.raw,col=col, las=2)
title(main="A. Raw data", ylab="log-cpm")
boxplot(cpm(dge.filter, log=T), col=col, las=2)
title(main="B. Normalized data")
```

```{r}
par(mfrow=c(3,1))
hist(aveLogCPM(dge))
hist(aveLogCPM(dge.test))
hist(aveLogCPM(dge.filter))
```

## Unsupervised clustering

```{r}
site <- as.factor(dge.filter$samples$site)
patient <- as.factor(dge.filter$samples$patient)
```

```{r message=FALSE, warning=FALSE}
par(mfrow=c(1,2))
# assign colour to different sites (normal vs poly)
col.site <- site
levels(col.site) <- brewer.pal(nlevels(col.site), "Set1")
col.site <- as.character(col.site)
# assign colour to different patients
col.patient <- dge.filter$samples$patient
levels(col.patient) <- brewer.pal(nlevels(col.patient), "Set3")
col.patient <- as.character(col.patient)

# plot with plotMDS - site
plotMDS(dge.filter, labels=site, col=col.site)
title(main="A. Sample site groups")

# plot with plotMDS - patient
plotMDS(dge.filter, labels=patient, col=col.patient)
title(main="B. Patient groups")
```

## Use Glimma to generate interactive html for the top PCAs

```{r}
library(Glimma)
glMDSPlot(lcpm.filter, labels=paste(site, patient, sep="_"), groups=dge.filter$samples[,6:7],
         launch=TRUE)
```

## DE analysis

### Create design matrix

```{r}
data.frame(sample=colnames(dge.filter), site, patient)
```

```{r}
design <- model.matrix(~0+patient+site)
colnames(design) <- gsub("site|patient", "", colnames(design))
rownames(design) <- colnames(dge.filter)
head(design)[,1:16]
```

### Dispersion estimation

```{r}
library(statmod)
```

```{r}
dge.filter <- estimateDisp(dge.filter, design, robust = T)
dge.filter$common.dispersion
```

```{r}
plotBCV(dge.filter, ylim=c(0,2))
```

```{r}
fit <- glmFit(dge.filter, design)
lrt <- glmLRT(fit)
topTags(lrt)
```

```{r}
o <- order(lrt$table$PValue)
cpm(dge.filter)[o[1:10], 1:6]
```

```{r}
summary(decideTests(lrt))
```

```{r}
par(mar=c(4,4,2,4))
plotMD(lrt,bg.pch=1)
abline(h=1, col="blue")
```

### Quasi-likelihood analysis

```{r}
fit_ql <- glmQLFit(dge.filter, design, robust = T)
head(fit_ql$coefficients)
```

```{r}
plotQLDisp(fit_ql)
```

### TSNE analysis

```{r}
tsne_out <- Rtsne(t(dge.filter$counts), perplexity = 5, dims=2, pca = T)
```

```{r}
plot(tsne_out$Y, col=site)
```

## Gene ontology analysis

```{r message=FALSE, warning=FALSE}
# load the GO annotation package
library(GO.db)
library(org.Hs.eg.db)
```

```{r}
# generate annotation 
annot <- select(org.Hs.eg.db, dge.filter$genes$genes, "SYMBOL")
```

```{r}
plot(scale(dge.filter$counts[rownames(dge.filter) == 'ENSG00000139800',]),
     scale(dge.filter$counts[rownames(dge.filter) == 'ENSG00000129451',]),
     pch=19,
     ylab='KLK10 (ENSG00000129451)',
     xlab='ZIC5 (ENSG00000139800)',
     col=site,
     xlim=c(-2,6),
     ylim=c(-2,6)
     )
```

```{r}
go <- goana(lrt)
```

```{r}
par(mfrow=c(1,3))
plotMDS(dge, labels=site, col=col.site)
plotMDS(dge.test, labels=site, col=col.site)
plotMDS(dge.filter, labels=site, col=col.site)
```
